cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,248 @@
+    }
+    return logFileOutputStream;
+  }
+
+  /**
+   * closes all the open avro files in the map
+   */
+  public void close() {
+    // close all the files in fileMap
+    // clear the map
+    for (LogFileOutputStream file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation(), e);
+      } finally {
+        Closeables.closeQuietly(file);
+      }
+    }
+    fileMap.clear();
+  }
+
+  private void closeAndDelete(LogFileOutputStream logFileOutputStream) {
+    try {
+      try {
+        logFileOutputStream.close();
+      } finally {
+        if (logFileOutputStream.getLocation().exists()) {
+          logFileOutputStream.getLocation().delete();
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while closing and deleting file {}", logFileOutputStream.getLocation(), e);
+    }
+  }
+
+  /**
+   * flushes the contents of all the open avro files
+   * @throws IOException
+   */
+  public void flush() throws IOException {
+    // perform flush on all the files in the fileMap
+    for (LogFileOutputStream file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation());
+        Closeables.closeQuietly(file);
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Represents an Avro file.
+   *
+   * Since there is no way to check the state of the underlying file on an exception,
+   * all methods of this class assume that the file state is bad on any exception and close the file.
+   */
+  class LogFileOutputStream implements Closeable {
+    private final Location location;
+    private FSDataOutputStream outputStream;
+    private DataFileWriter<GenericRecord> dataFileWriter;
+    private long createTime;
+
+    LogFileOutputStream(Location location) throws IOException {
+      this.location = location;
+      try {
+        this.outputStream = new FSDataOutputStream(location.getOutputStream(), null);
+        this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<GenericRecord>(schema));
+        this.dataFileWriter.create(schema, this.outputStream);
+        this.dataFileWriter.setSyncInterval(syncIntervalBytes);
+        this.createTime = System.currentTimeMillis();
+      } catch (Exception e) {
+        Closeables.closeQuietly(dataFileWriter);
+        Closeables.closeQuietly(outputStream);
+        throw new IOException("Exception while creating file " + location, e);
+      }
+    }
+
+    public Location getLocation() {
+      return location;
+    }
+
+    public void append(ILoggingEvent event) throws IOException {
+      try {
+        dataFileWriter.append(co.cask.cdap.logging.serialize.LoggingEvent.encode(schema, event));
+      } catch (Exception e) {
+        throw new IOException("Exception while appending to file " + location, e);
```

Comment:
This is not good as it always wrap the exception. Also, the `DataFileWriter.append()` has `throws IOException` only, why you need to catch and rethrow?